Manutenção de alunos

pedro-gandra opened the pull request:
# [Grupo 5] Manutenção de alunos (completa)

* Métodos de inserir, alterar, deletar, listar e consultar alunos
* Outras funcionalidades: aluno alterar sua própria senha/foto 
* Validação completa para todos as funcionalidades
* Login do aluno através de id e senha implementado

## Cartões

* [Back-end para inserir, alterar e remover (manutenção de alunos)] (https://trello.com/c/U8LueoTH/12-back-end-para-inserir-alterar-e-remover-manuten%C3%A7%C3%A3o-de-alunos)

* [Permissões para manutenção de alunos] (https://trello.com/c/7PqmWs8s/14-permiss%C3%B5es-para-manuten%C3%A7%C3%A3o-de-alunos)

pedro-gandra committed:
Manutenção de Alunos

AlunosRepository e servlets para inserção, alteração e deletação de alunos

pedro-gandra committed:
Merge remote-tracking branch 'origin/feature/diario/alunos' into feature/diario/alunos

pedro-gandra committed:
Validação e melhoria na deletação

Checa se o usuário é um ADMIN, caso tente executar funcionalidades exclusivas de ADMINS.
Antes de deletar um aluno verifica se esse tem uma matricula registrada com sua identificação, caso tenha, uma mensagem de erro dizendo que a matricula deve ser deletada antes do aluno é retornada

pedro-gandra committed:
Listagem e consulta de alunos

Adicionadas funcionalidades de listar e consultar alunos por id

pedro-gandra committed:
Aprimoramento da estrutura

Utilizando AlunoRepository para métodos de consultar e listar

pedro-gandra committed:
Melhoria na estrutura

O método de deletar aluno em AlunoRepository agora checa se existe uma matrícula vinculada aquele aluno.
Correção de um dos xmls de resposta

pedro-gandra committed:
Fix

Tinha esquecido de mudar uma condição usada para teste

pedro-gandra committed:
Merge remote-tracking branch 'origin/dev-s1' into feature/diario/alunos

pedro-gandra committed:
Servlets para o aluno alterar a própria senha/foto

Criação de dois servlets, um para o aluno alterar a própria senha e outro para alterar a própria foto

pedro-gandra committed:
Fix gitkraken bug

Commit havia atingido arquivo de outra branch, revertendo uma das mudanças feitas

pedro-gandra committed:
Correção de servlets

Adiciona checagem de aluno para conferir se o aluno está tentando alterar dados de OUTRO aluno, criação do servlet AlterarFoto que havia sido deletado devido à bug

pedro-gandra committed:
Correção servlet de deletar alunos

Corrigindo um bug relacionado ao retorno xml

pedro-gandra committed:
Criação de funcionalidade para o login do aluno

Criado servlet para logar o aluno, tendo como parâmetros id e senha, além de correção no tratamento de algumas exceções

pedro-gandra committed:
Formatação

Correção da formatação e remoção de coisas inúteis

LiviaLelis commented:
Desculpa

LiviaLelis commented on app/src/java/diario/alunos/repository/AlunosRepository.java:
```diff
@@ -0,0 +1,251 @@
+import java.util.logging.Logger;
+import javax.servlet.http.HttpServletResponse;
+import utils.Hasher;
+import static utils.Hasher.hash;
```
\\\\\\ gl, hf

LiviaLelis commented on app/src/java/diario/alunos/repository/AlunosRepository.java:
```diff
@@ -0,0 +1,251 @@
+
+	}
+
+	public boolean alterarAlunos(String id, String nome, String email, String senha, String sexo, String nascimento, String logradouro, String numero, String complemento, String bairro, String cidade, String cep, String uf, String foto  ) throws NumberFormatException, SQLException, NoSuchAlgorithmException, ParseException {
```
Query vulnerável

LiviaLelis commented on app/src/java/diario/alunos/repository/AlunosRepository.java:
```diff
@@ -0,0 +1,251 @@
+		}
+
+
+		query += " WHERE `id` = "+id+"";
```
Query vulnerável

LiviaLelis commented on app/src/java/diario/alunos/repository/AlunosRepository.java:
```diff
@@ -0,0 +1,251 @@
+		ps.setInt(1, idParsed);
+		ResultSet rs = ps.executeQuery();
+		while(rs.next()) {
+			xml += viewConsulta.XMLAlunoCompleto(rs.getInt("id"), rs.getString("nome"),  rs.getString("email"), rs.getString("sexo"), rs.getDate("nascimento"), rs.getString("logradouro"), rs.getInt("numero"), rs.getString("complemento"), rs.getString("bairro"),rs.getString("cidade"), rs.getInt("cep"),rs.getString("uf"), rs.getString("foto"));
```
Quebra a linha

LiviaLelis commented on app/src/java/diario/alunos/servlets/InserirAlunos.java:
```diff
@@ -0,0 +1,92 @@
+				Logger.getLogger(InserirAlunos.class.getName()).log(Level.SEVERE, null, ex);
+			} 
+		} else {
+			response.setStatus(401);
```
403

LiviaLelis reviewed (changes requested):
Corrigir as queries, depois testarei a qualidade e estará sujeito a cprreção

PedroCesarMesquita commented on app/src/java/diario/alunos/repository/AlunosRepository.java:
```diff
@@ -0,0 +1,251 @@
+
+	public boolean inserirAlunos(String id, String nome, String email, String senha, String sexo, String nascimento, String logradouro, String numero, String complemento, String bairro, String cidade, String cep, String uf, String foto  ) throws NumberFormatException, SQLException, NoSuchAlgorithmException, ParseException, InvalidKeySpecException {
+
+
```
Remover linhas vazias.

pedro-gandra committed:
Correção de erros

- Melhoria na query
-  Mudanças de status para operação não autorizada
- Organização melhor do código
- Correção de alguns outros erros não apontados pelo revisor da PR

LiviaLelis commented on app/src/java/diario/alunos/servlets/AlterarAlunos.java:
```diff
@@ -0,0 +1,96 @@
+			} catch (RenderException e) {
+				throw new ServletException(e);
+			} catch (NoSuchAlgorithmException | ParseException ex) {
+				out.println("<erro><mensagem>Erro severo</mensagem></erro>");
```
Modificar o status code

LiviaLelis reviewed (changes requested):
Modificar a linha marcada.
_Observações:_ o código está inconsistente em algumas partes, alguns métodos estão confusos (você sabe muito bem qual é) e serão de difícil manutenção ou utilização por outra pessoa.

pedro-gandra committed:
Modificando status code

Modificação de status code em caso de erros específicos em um dos servlets

ppassinho commented on app/src/java/diario/alunos/repository/AlunosRepository.java:
```diff
@@ -0,0 +1,350 @@
+	}
+
+
+	public String deletarAlunos(String id) throws SQLException{
```
## Duas sugestões para retorno de erros:

* Alterar tipo do método para `void` e lançar exceções
  *ou*
* Alterar tipo do método para `boolean` e retornar `sucesso != 0`

---

## Sugestão para parâmetro `id`

Alterar tipo para `long`

* Usar um tipo numérico para o `id` é mais prático do que usar o tipo `String`
* O `id` de um aluno corresponde ao `CPF` do mesmo, o qual excede o limite de inteiros de 32 bits

ppassinho commented on app/src/java/diario/alunos/repository/AlunosRepository.java:
```diff
@@ -0,0 +1,350 @@
+		return "erro";
+	}
+
+	public boolean inserirAlunos(String id, String nome, String email, String senha, String sexo, String nascimento, String logradouro, String numero, String complemento, String bairro, String cidade, String cep, String uf, String foto  ) throws NumberFormatException, SQLException, NoSuchAlgorithmException, ParseException, InvalidKeySpecException {
```
## Duas sugestões para métodos `inserirAlunos` e `alterarAlunos`

* Utilizar um model `Aluno` que contenha todas informações de um aluno
  *ou*
* Utilizar um `Map` contendo todas informações de um aluno

pmesq reviewed (changes requested):
# Revisão

## Alterar tipo do id do aluno para `long`

O `id` de um aluno é o seu CPF, o qual não cabe em um inteiro de 32 bits (`int`), mas sim em um inteiro de 64 bits (`long`).

---

## [Sugestão] Corrigir identação

Dica: utilizar a opção de formatar do NetBeans
`Menu superior` -> `Código-fonte` -> `Formatar`

---

## [Sugestão] Remover ou corrigir documentação de classes

Exemplo:

```java
/**
 * Descrição da classe
 * @author pedro-gandra
 */
```

---

## [Sugestão] Utilizar models

Criar um model `Aluno`, contendo todas informações do mesmo, e utilizá-lo nos códigos.

---

## [Sugestão] Utilizar views

Criar classes de view para construção dos XMLs.

pedro-gandra committed:
Merge remote-tracking branch 'origin/feature/diario/alunos' into feature/diario/alunos

pedro-gandra committed:
Correção de erros

- Correção de formatação
- Remoção de comentários indevidos
- Alteração do tamanho do inteiro na identificação do aluno

pedro-gandra committed:
Alteração de tamanho de inteiro no login de aluno

- Usando Long agora, tanto no DiarioAutenticador quanto no método específico de logar aluno

pedro-gandra committed:
Correção de status code

- Corrigindo status code de SQL Exception de um servlet
- Corrigindo status code de erro de permissão de alguns servlets
- Corrigindo retorno xml de um servlet

pedro-gandra committed:
Auto stash before merge of "feature/diario/alunos" and "origin/feature/diario/alunos"

pedro-gandra committed:
Merge remote-tracking branch 'origin/dev-s1' into feature/diario/alunos

pedro-gandra committed:
Corrigindo problemas com CPF

Agora o usuário sempre visualiza o cpf corretamente, mesmo que o banco de dados tenha "comido" os zeros à esquerda

pedro-gandra committed:
Fix

Remoção de condição usada para teste

ppassinho commented on app/src/java/utils/autenticador/DiarioAutenticador.java:
```diff
@@ -14,7 +14,7 @@ public DiarioAutenticador(HttpServletRequest request, HttpServletResponse respon
 	}
 
-	public void logar(Integer identididade, DiarioCargos cargo, boolean manter){
+	public void logar(Long identididade, DiarioCargos cargo, boolean manter){
```
[Essa PR](https://github.com/INF2-2019/servidor/pull/62), teoricamente, já corrige esses problemas, então pode haver conflitos caso ambas PRs sejam aprovadas.

LiviaLelis replied:
Recomendo aprovar a que eu fiz e resolver os conflitos localmente

ppassinho commented on app/src/java/diario/alunos/servlets/LogarAluno.java:
```diff
@@ -0,0 +1,53 @@
+		} catch (SQLException ex) {
+			out.println("<erro><mensagem>Usuario inexistente</mensagem></erro>");
+		} catch (NoSuchAlgorithmException ex) {
+			Logger.getLogger(LogarAluno.class.getName()).log(Level.SEVERE, null, ex);
```
Retornar XML de erro

ppassinho commented on app/src/java/diario/alunos/servlets/LogarAluno.java:
```diff
@@ -0,0 +1,53 @@
+		} catch (NoSuchAlgorithmException ex) {
+			Logger.getLogger(LogarAluno.class.getName()).log(Level.SEVERE, null, ex);
+		} catch (InvalidKeySpecException ex) {
+			Logger.getLogger(LogarAluno.class.getName()).log(Level.SEVERE, null, ex);
```
Retornar XML de erro
